### PR TITLE
for pmd 7 pmd-java and pmd-ant are required

### PR DESCRIPTION
**Pull request by madmuffin1**
Related to #24502
The default rules still need some adressing.

### Context
PMD 7.0.0 introduces major improvements, give gradle users the chance to upgrade

### Contributor Checklist
- [X] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [X] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [X] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [X] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [X] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [X] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [X] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java [hunk 1]
```diff
@@ -144,11 +145,11 @@ private List<String> ruleSetsConvention(PmdExtension extension) {
...
-        configuration.defaultDependencies(dependencies -> {
-                VersionNumber version = VersionNumber.parse(extension.getToolVersion());
-                String dependency = calculateDefaultDependencyNotation(version);
-                dependencies.add(project.getDependencies().create(dependency));
-            }
+        configuration.defaultDependencies(dependencies ->
+            calculateDefaultDependencyNotation(extension.getToolVersion())
+            .stream()
+            .map(project.getDependencies()::create)
+            .forEach(dependencies::add)
```
Comment (jvandort) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java:152]:
```suggestion
        configuration.defaultDependencies(dependencies ->
            calculateDefaultDependencyNotation(extension.getToolVersion())
                .stream()
                .map(project.getDependencies()::create)
                .forEach(dependencies::add)
```

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java [hunk 2]
```diff
@@ -192,13 +193,21 @@ private void configureToolchains(Pmd task) {
...
+        final VersionNumber toolVersion = VersionNumber.parse(versionString);
         if (toolVersion.compareTo(VersionNumber.version(5)) < 0) {
-            return "pmd:pmd:" + extension.getToolVersion();
+            return Set.of("pmd:pmd:" + versionString);
         } else if (toolVersion.compareTo(VersionNumber.parse("5.2.0")) < 0) {
-            return "net.sourceforge.pmd:pmd:" + extension.getToolVersion();
+            return Set.of("net.sourceforge.pmd:pmd:" + versionString);
+        } else if (toolVersion.compareTo(VersionNumber.version(7)) >= 0) {
+            // starting from version 7, PMD is split into multiple modules
+            return Set.of(
```
Comment (jvandort) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java:204]: We need to support Java 8. `Set.of(...)` will not work

**Review by jvandort — CHANGES_REQUESTED** [review]

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java (hunk 2 above)
Comment (madmuffin1) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java:204]: no probs, I will migrate this

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java [hunk 3]
```diff
@@ -192,13 +193,21 @@ private void configureToolchains(Pmd task) {
...
+                "net.sourceforge.pmd:pmd-java:" + versionString,
+                "net.sourceforge.pmd:pmd-ant:" + versionString
+
+            );
         }
-        return "net.sourceforge.pmd:pmd-java:" + extension.getToolVersion();
+        return Set.of("net.sourceforge.pmd:pmd-java:" + versionString);
```
[the remark this answers is not in the document]
Comment (madmuffin1) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java:210]: you're right, I kept the still default in the fallthrough, but it is not very intuitive

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java (hunk 2 above)
Comment (madmuffin1) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java]: should be JRE8 compatible now.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (madmuffin1) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java]: reordered.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java [hunk 4]
```diff
@@ -192,13 +194,21 @@ private void configureToolchains(Pmd task) {
         });
     }
 
-    private String calculateDefaultDependencyNotation(VersionNumber toolVersion) {
+    static Set<String> calculateDefaultDependencyNotation(final String versionString) {
```
Comment (jvandort) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java:197]: This needs a `@VisibleForTesting`

**Review by jvandort — CHANGES_REQUESTED** [review]
The implementation LGTM but we should also add integration test support for PMD 7. That would include updating `org.gradle.quality.integtest.fixtures.PmdCoverage` to add an entry for 7.0.0-rc2 and verify all tests succeed. If you make the change to `PmdCoverage` I can start the tests on CI

**Comment by jvandort** [comment]
> The implementation LGTM but we should also add integration test support for PMD 7. That would include updating `org.gradle.quality.integtest.fixtures.PmdCoverage` to add an entry for 7.0.0-rc2 and verify all tests succeed. If you make the change to `PmdCoverage` I can trigger the tests on CI to verify the coverage

You mentioned this only partially fixes #24502. In that case we can hold off on the integration test updates.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java (hunk 4 above)
Comment (madmuffin1) [reply subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdPlugin.java:199]: done

**Comment by jvandort** [comment]
@bot-gradle test qfl

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**